Lab3

eggorrus opened the pull request.

eggorrus committed:
Написан CycleList.h

eggorrus committed:
lab3 v2

написан "каркас" работы, но нужно доработать саму программу лабораторной и написать xml

eggorrus committed:
lab3 v3

Добавлена xml, исправлены недочеты в коде лабораторной работы

ichepkov2401 commented on 2sem/common/CycleList.h:
```diff
@@ -0,0 +1,194 @@
+class CycleList
+{
+private:
+	Node<T>* beginning;
```
Начало списка выражается через следующий элемент последнего элемента

eggorrus replied:
реализовано

ichepkov2401 commented on 2sem/common/CycleList.h:
```diff
@@ -0,0 +1,194 @@
+		}
+		else
+		{
+			Node<T>* current = beginning;
```
Оптимизировать для прямого прохода по списку

eggorrus replied:
реализовано

ichepkov2401 commented on 2sem/common/CycleList.h:
```diff
@@ -0,0 +1,194 @@
+		else
+		{
+			end->next = newNode;
+			end = newNode;
```
Не реализована цикличность списка

eggorrus replied:
исправлено

ichepkov2401 commented on 2sem/lab3/lab3.cpp:
```diff
@@ -1,20 +1,169 @@
+void insertOneBeforeNegatives(CycleList<int>* list, int sizeOfList)
+{
+	Node<int>* current = list->returnBeginning();
+	for (int i = 0; i < sizeOfList; i++)
```
```suggestion
	for (int i = 0; i < list.count(); i++)
```

eggorrus replied:
исправлено

ichepkov2401 commented on 2sem/lab3/lab3.cpp:
```diff
@@ -1,20 +1,169 @@
+void removeNegatives(CycleList<int>* list, int sizeOfList)
+{
+	Node<int>* current = list->returnBeginning();
+	for (int i = 0; i < sizeOfList; i++)
```
```suggestion
	for (int i = 0; i < list.count(); i++)
```

eggorrus replied:
исправлено

ichepkov2401 commented on 2sem/common/CycleList.h:
```diff
@@ -0,0 +1,194 @@
+	/// Позволяет обращаться к началу списка
+	/// </summary>
+	/// <returns> Указатель на первый элемент списка </returns>
+	Node<T>* returnBeginning()
```
Не должно торчать наружу

eggorrus replied:
можно уточнить, о чем речь?

ichepkov2401 replied:
Тот кто пользуется вашим списком ничего не должен знать о том, как он устроен внутри. Поэтому никакие методы возвращающие узлы не должны быть публичными
![image](https://github.com/user-attachments/assets/63a6d2de-34ad-495a-9f49-3c9921b44a9c)
![image](https://github.com/user-attachments/assets/0abd8418-d588-4881-936e-1332d8f5ccd6)

ichepkov2401 commented on 2sem/lab3/lab3.cpp:
```diff
@@ -1,20 +1,169 @@
+void outputList(CycleList<int>* list)
 {
-    std::cout << "Hello World!\n";
+	int sizeOfList = list->count();
```
Может просто list->count() использовать?

eggorrus replied:
исправлено на этот вариант

eggorrus committed:
lab3 v4

Метод count теперь принимает в качестве параметра не ссылку а значение value

eggorrus committed:
labe v5

1. список сделан цикличным, каким он и должен был быть с самого начала
2. Оптимизированы функции и методы

eggorrus committed:
lab3 v6

исправлена ошибка в методе insert (разыменование нулевого указателя)

ichepkov2401 commented on 2sem/common/CycleList.h:
```diff
@@ -0,0 +1,216 @@
+		}
+		else
+		{
+			Node<T>* current = end->next;
```
Оптимизировать для прямого прохода по коллекции

ichepkov2401 commented on 2sem/common/CycleList.h:
```diff
@@ -0,0 +1,216 @@
+				throw std::out_of_range("Invalid index for empty list");
+			}
+			end = newNode;
+			end->next = end;
```
Дублирование кода add()

ichepkov2401 commented on 2sem/common/CycleList.h:
```diff
@@ -0,0 +1,216 @@
+		if (counter == 0)
+		{
+			if (index != 0)
+			{
```
Недостижимы код

ichepkov2401 commented on 2sem/common/CycleList.h:
```diff
@@ -0,0 +1,216 @@
+			end->next = newNode;
+		}
+		else if (index == counter)
+		{
```
Дублирование кода add()

eggorrus committed:
lab3 v7

Переработаны методы, связанные со вставкой элемента в список, убран метод returnBeginning()

eggorrus committed:
lab3 v8

исправлен метод insert

eggorrus committed:
optimized list

добавлен указатель currentNode и int currentCount, переработан метод вставки единиц (возник конфликт между методами оператора и Insert)

eggorrus commented:
Вынужден закрыть пулл реквест, чтобы слить ветку для курсовой.
